### PR TITLE
193 - Performance trace

### PR DESCRIPTION
**Pull request by maeisabelle**
Added performance trace in quick-start (using aop), DataHub core libraries called in quick start (using a utility), and xquery files inside ml-modules/(services|transforms) (using a module). Kindly review. This is to address #193.

### CODE REVIEW

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/perflog-lib.xqy [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+:)
+xquery version "1.0-ml";
+
+module namespace perflog = "http://marklogic.com/data-hub/perflog-lib";
+
+(:~
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/perflog-lib.xqy]: Your comment doesn't quite match up to the function

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/perflog-lib.xqy [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ : determine if a module exists or not
+ :
+ : @param $uri - the uri of the module to check
+ : @return - true if exists, false otherwise
+ :)
+declare function perflog:logit($what-is-it, $x)
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/perflog-lib.xqy]: You weren't on the call when I talked about this code. I mentioned that it was an example and that you all shouldn't use the exact names as I wrote it in about 45 seconds. I'd like to see the variable names make a little more sense.

Also I'd prefer to see some sort of greppable thing in the log message. Maybe something like:

xdmp:log("PERFORMANCE: " || $what-is-it || " took " || $time)

You should also look at the log file output to make sure it makes sense. I'm pretty sure $time is something like "PT0.00001S" and doesn't need the ms at the end.

File: marklogic-data-hub/src/main/resources/ml-modules/services/collector.xqy [hunk 3]
```diff
@@ -31,9 +34,11 @@ declare option xdmp:mapping "false";
 declare function service:build-response($items as item()*)
   as document-node()
 {
-  document {
-    json:to-array($items) ! xdmp:to-json(.)
-  }
+  perflog:logit('Collector.buildResponse',function() {
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/services/collector.xqy]: I think wrapping this function in the perflog is a bit overkill. It's already covered by the functions below. The idea is to get a sense of how long a transaction is taking. Thus, I'd prefer to see 1 log per transaction.

File: marklogic-data-hub/src/main/resources/ml-modules/services/collector.xqy [hunk 4]
```diff
@@ -46,12 +51,14 @@ declare function post(
   $input   as document-node()*
   ) as document-node()*
 {
-  debug:dump-env(),
+  perflog:logit('Collector.post',function() {
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/services/collector.xqy]: I'd like to see a better name logged. at the minimum use the : vs the .

perflog:logit('collector:post')

but maybe

perflog:logit('GET => /v1/resources/collector')

I'm not sure what works best. I just want it to be pretty easy to track down the offending code if I find a performance issue.

File: marklogic-data-hub/src/main/resources/ml-modules/services/debug.xqy [hunk 5]
```diff
@@ -19,6 +19,9 @@ module namespace service = "http://marklogic.com/rest-api/resource/debug";
 
 import module namespace debug = "http://marklogic.com/data-hub/debug-lib"
   at "/com.marklogic.hub/lib/debug-lib.xqy";
+
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/services/debug.xqy]: I don't think it makes sense to perf-log this file. It's so basic. Please back this file out of the PR.

File: marklogic-data-hub/src/main/resources/ml-modules/services/tracing.xqy [hunk 6]
```diff
@@ -23,6 +23,9 @@ import module namespace debug = "http://marklogic.com/data-hub/debug-lib"
 import module namespace trace = "http://marklogic.com/data-hub/trace"
   at "/com.marklogic.hub/lib/trace-lib.xqy";
 
+import module namespace perflog = "http://marklogic.com/data-hub/perflog-lib"
```
Comment (paxtonhare) [inline marklogic-data-hub/src/main/resources/ml-modules/services/tracing.xqy]: This module doesn't need performance tracing either. Please back it out of the PR.

File: quick-start/build.gradle [hunk 7]
```diff
@@ -42,6 +42,7 @@ dependencies {
         exclude group: 'com.google.guava', module: 'guava'
     }
     compile 'com.google.guava:guava:19.0'
+    compile 'org.springframework.boot:spring-boot-starter-aop:1.3.3.RELEASE'
```
Comment (paxtonhare) [inline quick-start/build.gradle]: Any clue how much larger this makes the quick-start.jar?

Comment (maeisabelle) [inline quick-start/build.gradle]: Hi @paxtonhare,

I compared the jar file created in master and in my branch, from 67.134229 MB, it became 69.004115 MB (so almost 2 MB). 

By the way, I noticed you merged my branch to your branch 'develop' and saw your changes. Would I still continue with the changes you requested?

**Comment by paxtonhare** [comment]
I manually merged this.

**Pull request closed without merging**